### DISCUSSION TITLE
Proposal: new node transforms API to replace `quill.on('text-change', fn)`

### DISCUSSION THREAD

**Discussion by lawvs** [toeverything/blocksuite#1012] [Ideas]
The quill on change API has the potential risk of breaking undo/redo or causing UI and YDoc to be inconsistent. (see https://github.com/toeverything/blocksuite/issues/415) 

I propose to create a new Text API to solve the problem.

```ts
Text.onUpdate((delta, context) => {
  // Do something
  // ...
  return newDelta;
})
```

Related code at
https://github.com/toeverything/blocksuite/blob/92c1d9392f4107bdad8d803b2f6ce4891fd25f87/packages/store/src/text-adapter.ts#L374-L400

Related to https://github.com/toeverything/blocksuite/issues/415 https://github.com/toeverything/AFFiNE/issues/632

Reference to https://lexical.dev/docs/concepts/transforms

**Comment by Flrande** [comment]
Do you think we can fix it by replacing quill with virgo?

**Reply by lawvs** [reply]
Virgo has no transform API now, but it is also compatible with the proposal API.

**Reply by Flrande** [reply]
Virgo has no transform API by design, because it directly implement the synchronization between dom and Y.Text, which mean that Y.Text is the single truth. (It requires Y.Text data structure is expected, that's why I consider add zod to validate insert data.)
